pkg/pipeline/stream/base.go

```diff
@@ -167,6 +175,22 @@ func (n *pubNodeBase) Trigger(
 	return trigger, cleanup, nil
 }
 
+// InjectMessage can be used to inject a message into the message stream. This
+// is used to inject control messages like the last position message when
+// stopping a source connector. It is a bit hacky, but it doesn't require us to
+// create a separate channel for signals which makes it performant and easiest
+// to implement.
+func (n *pubNodeBase) InjectMessage(ctx context.Context, message *Message) error {
```

Comment:
Good idea, but instead of adding a check I changed the function signature to only allow control messages in the first place:
```
func (n *pubNodeBase) InjectControlMessage(ctx context.Context, msgType ControlMessageType) error {
```

Reply:
> What if we have a method called `StopGracefully` or smth along those lines

This is a base node so it has no idea what "stopping a node" means, the actual implementation needs to make that decision.

Reply:
Yeah, I just recalled that it's only the source node for which you can request a stop, all others are "stoppable" in the sense that they _can_ be stopped, but not through a direct request.